File: projecteuler.fsharp/ProjectEulerTasks/Problem083.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open System.IO
open CommonLib

// In the 5 by 5 matrix (more see here: https://projecteuler.net/problem=83), the minimal path sum from the top left to the bottom right, by moving left, right, up, and down, is equal to 2297.
// Find the minimal path sum, in problem_083.dat, a 31K text file containing a 80 by 80 matrix, from the top left to the bottom right by moving left, right, up, and down.

[<TestFixture>]
type Problem083() =

    let solveImpl (dataFilename: string) =
        let grid = File.ReadAllLines(Path.Combine("Data", dataFilename)) |> Seq.map (fun line -> line.Split(',') |> Seq.map int |> Seq.toArray) |> Seq.toArray
        let rowMax = grid.Length
        let columnMax = grid.[0].Length
        let  generateNextPoints (point: GridPoint) =
            match point with
            | _ when (point.Row = rowMax) && (point.Column = columnMax) -> []
            | _ when (point.Row = 1) && (point.Column = 1) -> [GridPoint(1, 2); GridPoint(2, 1)]
            | _ when (point.Row = 1) && (point.Column = columnMax) -> [GridPoint(2, columnMax); GridPoint(1, columnMax - 1)]
            | _ when (point.Row = rowMax) && (point.Column = 1) -> [GridPoint(rowMax - 1, 1); GridPoint(rowMax, 2)]
            | _ when (point.Row = 1) -> [GridPoint(2, point.Column); GridPoint(1, point.Column - 1); GridPoint(1, point.Column + 1)]
            | _ when (point.Row = rowMax) -> [GridPoint(rowMax - 1, point.Column); GridPoint(rowMax, point.Column - 1); GridPoint(rowMax, point.Column + 1)]
            | _ when (point.Column = 1) -> [GridPoint(point.Row - 1, 1); GridPoint(point.Row + 1, 1); GridPoint(point.Row, 2)]
            | _ when (point.Column = columnMax) -> [GridPoint(point.Row - 1, columnMax); GridPoint(point.Row + 1, columnMax); GridPoint(point.Row, columnMax - 1)]
            | _ -> [GridPoint(point.Row - 1, point.Column); GridPoint(point.Row + 1, point.Column); GridPoint(point.Row, point.Column - 1); GridPoint(point.Row, point.Column + 1)]
        let behavior = { new IGridPathBehavior<int> with
                             member this.GenerateNextValue(accValue: int, pointValue: int) = accValue + pointValue
                             member this.CompareValues(leftValue: int, rightValue: int) = rightValue - leftValue
                             member this.GenerateNextPoint (point: GridPoint) = point |> generateNextPoints
                       }
        let pathFactory = GridPathFactory<int>(behavior)
        let result = pathFactory.Create(grid, [GridPoint(1, 1)], [GridPoint(rowMax, columnMax)])
        result.Value

    [<TestCase("problem_083.dat", 425185, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(dataFilename: string, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, dataFilename)

File: projecteuler.erlang/project_euler/src/problem_083.erl
%% @author std-string

%% In the 5 by 5 matrix (in problem_083_example.dat), the minimal path sum from the top left to the bottom right, by moving left, right, up, and down, is equal to 2297.
%% Find the minimal path sum, in problem_083.dat, a 31K text file containing a 80 by 80 matrix, from the top left to the bottom right by moving left, right, up, and down.

-module(problem_083).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

%% ====================================================================
%% API functions
%% ====================================================================

-spec get_check_data() -> [{Input :: term(), Output :: term()}].
get_check_data() -> [{"problem_083_example.dat", 2297}, {"problem_083.dat", 425185}].

-spec prepare_data(ModuleSourceDir :: string(), Input :: term()) -> term().
prepare_data(ModuleSourceDir, Filename) ->
    load_utils:read_number_table(filename:join(ModuleSourceDir, Filename), ",").

-spec solve(PreparedInput :: term()) -> term().
solve(GridData) ->
    Grid = grid:copy(GridData),
    InitPoints = [{1, 1}],
    ResultPoints = [{grid:get_row_count(Grid), grid:get_column_count(Grid)}],
    ValueBuilder = fun(AccValue, PointValue) -> AccValue + PointValue end,
    ValueComparator = fun compare:compare_asc/2,
    {Value, _Path} = grid_path:search(Grid, InitPoints, ResultPoints, ValueBuilder, ValueComparator, fun get_next_points/3),
    Value.

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec get_next_points(Point :: grid:point_type(), RowMax :: grid:row_type(), ColumnMax :: grid:column_type()) -> [grid:point_type()].
get_next_points({RowMax, ColumnMax}, RowMax, ColumnMax) -> [];
get_next_points({1, 1}, _RowMax, _ColumnMax) -> [{1, 2}, {2, 1}];
get_next_points({1, ColumnMax}, _RowMax, ColumnMax) -> [{2, ColumnMax}, {1, ColumnMax - 1}];
get_next_points({RowMax, 1}, RowMax, _ColumnMax) -> [{RowMax - 1, 1}, {RowMax, 2}];
get_next_points({1, Column}, _RowMax, _ColumnMax) -> [{2, Column}, {1, Column - 1}, {1, Column + 1}];
get_next_points({RowMax, Column}, RowMax, _ColumnMax) -> [{RowMax - 1, Column}, {RowMax, Column - 1}, {RowMax, Column + 1}];
get_next_points({Row, 1}, _RowMax, _ColumnMax) -> [{Row - 1, 1}, {Row + 1, 1}, {Row, 2}];
get_next_points({Row, ColumnMax}, _RowMax, ColumnMax) -> [{Row - 1, ColumnMax}, {Row + 1, ColumnMax}, {Row, ColumnMax - 1}];
get_next_points({Row, Column}, _RowMax, _ColumnMax) -> [{Row - 1, Column}, {Row + 1, Column}, {Row, Column - 1}, {Row, Column + 1}].